### PR TITLE
Fix: Protect against potential nullptr dereference

### LINKED ISSUE #4738: Compiler warning: potential null pointer dereference
### What is the issue?

```
src/transmission/libtransmission/torrent.cc: In member function ‘std::string_view tr_torrent::primaryMimeType() const’:
src/transmission/libtransmission/torrent.cc:2320:16: warning: potential null pointer dereference [-Wnull-dereference]
 2320 |     return it->first;
      |                ^~~~~
src/transmission/libtransmission/torrent.cc:2320:16: warning: potential null pointer dereference [-Wnull-dereference]

```

```
In member function ‘tr_tracker* tr_tier::currentTracker()’,
    inlined from ‘tr_tracker* tr_tier::useNextTracker()’ at src/transmission/libtransmission/announcer.cc:419:30,
    inlined from ‘void {anonymous}::on_scrape_done_helpers::on_scrape_error(const tr_session*, tr_tier*, const char*)’ at src/transmission/libtransmission/announcer.cc:1245:43:
src/transmission/libtransmission/announcer.cc:357:20: warning: potential null pointer dereference [-Wnull-dereference]
  357 |             return nullptr;
      |                    ^~~~~~~
src/transmission/libtransmission/announcer.cc: In function ‘void {anonymous}::multiscrape(tr_announcer_impl*, const std::vector<tr_tier*>&)’:
src/transmission/libtransmission/announcer.cc:1404:65: warning: potential null pointer dereference [-Wnull-dereference]
 1404 |         auto const* const scrape_info = tier->currentTracker()->scrape_info;
      |   
```

### Which application of Transmission?

None

### Which version of Transmission?

4.0.0

### PR DESCRIPTION
**Pull request by Berbe**
Fixes #4738.

Notes: Fixed nullptr dereference compiler warning.

### CODE REVIEW

**Review by ckerr — CHANGES_REQUESTED** [review]
LGTM, but let's add a safeguard in case assertions are disabled as well:

File: libtransmission/announcer.cc [hunk 1]
```diff
@@ -1401,7 +1401,10 @@ void multiscrape(tr_announcer_impl* announcer, std::vector<tr_tier*> const& tier
     // batch as many info_hashes into a request as we can
     for (auto* tier : tiers)
     {
-        auto const* const scrape_info = tier->currentTracker()->scrape_info;
+        auto const* const current_tracker = tier->currentTracker();
+        TR_ASSERT(scrape_info != nullptr);
```
Comment (ckerr) [inline libtransmission/announcer.cc:1405]: Did you mean for this to be `current_tracker != nullptr`?

```suggestion
        auto const* const current_tracker = tier->currentTracker();
        TR_ASSERT(current_tracker != nullptr);
        if (current_tracker == nullptr)
        {
                    continue;
        }
```

Comment (ckerr) [inline libtransmission/announcer.cc:1405]: Did you mean for this to be `current_tracker != nullptr`?

```suggestion
        auto const* const current_tracker = tier->currentTracker();
        TR_ASSERT(current_tracker != nullptr);
        if (current_tracker == nullptr)
        {
            continue;
        }
```

Comment (Berbe) [reply libtransmission/announcer.cc:1405]: > Did you mean for this to be `current_tracker != nullptr`?

How copy-pasting can make one fail on a couple lines change… 😏 
I will merge your suggestion and push

Comment (Berbe) [reply libtransmission/announcer.cc]: I also added the same check from the immediately following, pre-existing assertion on `scrape_info`

**Comment by Berbe** [comment]
I got a question which is out of the scope of this specific PR, but is about the coding style.

I noticed `auto const* const` constructs were mainly used (although typing looks inconsistent throughout this single file) when I was rather expecting `const auto* const` ones: it's the `auto` type which is a pointer. I find the latter more readable, but I'm not experienced much.
What is the narrative behind the former construct?

**Comment by Coeur** [comment]
@Berbe It's right-to-left reading of the types in C: "a constant pointer to a constant auto-inferred type"

**Comment by ckerr** [comment]
> What is the narrative behind the former construct?

It's "east const." A couple of writeups: [1](https://isocpp.org/wiki/faq/const-correctness#const-ref-alt), [2](http://slashslash.info/eastconst/)

**Review by ckerr — COMMENTED** [review]
Thanks for the fix!

If you don't mind, I have one minor nit: I'd move the `bool found = false;` declaration to after the second block, so that the seven `current_tracker` lines and seven `scrape_info` lines have symmetry with each other.

Other than that, LGTM

**Comment by Berbe** [comment]
@ckerr Wilco!

On a side matter, I encountered these warnings building `4.0.0`. Since this is a code-clean job not changing any feature, maybe this fix could qualify for a patch release `4.0.1` if you ever make one?
I notice this project is not used to those. Or maybe are you versed in the "start from head" philosophy?

**Comment by Berbe** [comment]
> @Berbe It's right-to-left reading of the types in C: "a constant pointer to a constant auto-inferred type"

Thanks for the answer!
However, I find it odd the asterisk is attached to `const`. How about `auto const * const`?
Or shall this be read like `(auto const)* const`, from your English sentence?

**Comment by Berbe** [comment]
> > What is the narrative behind the former construct?
> 
> It's "east const." A couple of writeups: [1](https://isocpp.org/wiki/faq/const-correctness#const-ref-alt), [2](http://slashslash.info/eastconst/)

Oh I missed this comment. Thanks

**Review by ckerr — APPROVED** [review]

**Comment by ckerr** [comment]
> Since this is a code-clean job not changing any feature, maybe this fix could qualify for a patch release 4.0.1 if you ever make one?

Agreed, this will be in 4.0.1.

**Comment by Coeur** [comment]
> However, I find it odd the asterisk is attached to `const`. How about `auto const * const`?

The debate is as old as tabs versus spaces and predates my own birth. The style adopted by Transmission is opposite as the one adopted by Apple for instance. So don't ask me what I prefer, ah ah, I'm forced to follow the style that is imposed by this 18-years old project. :)
And since maybe 33% of the lines of code of the project would need to be rewritten to change the style, I'm not going to give pro or con, I'll just follow the flow about that. ;)

**Comment by ckerr** [comment]
I don't have strong opinions about any indentation style but I'm not touching `.clang-format` until 5.0.0 :laughing:

The idea of reformatting everything in a patch release makes me queasy

**Comment by ckerr** [comment]
I don't have strong opinions about any indentation style but I'm not touching `.clang-format` until 5.0.0 :laughing:

Style changes should be made very very rarely because they break backporting patches to other branches, which we'll have to do for every fix as soon as we make the 4.0.x branch

**Pull request merged**